opt/norm: add UseSwapMutation normalization rule

michae2 opened the pull request:
Add a new normalization rule which converts Update and Delete expressions to UpdateSwap and DeleteSwap, respectively.

With this normalization rule in place, UpdateSwap and DeleteSwap should be available for use by any qualifying UPDATE and DELETE statements.

See individual commits for details and release note.

Fixes: #144503
Fixes: #85328

Release note: None

michae2 committed:
opt: add swap flag to MutationPrivate

Inside the optimizer we can simply use a boolean flag to distinguish
between update and delete, and update swap and delete swap,
respectively. Add this flag to MutationPrivate and use it in execbuild
to construct the swap nodes.

Informs: #144503

Release note: None

michae2 committed:
opt/norm: add UseSwapMutation normalization rule

Add a new normalization rule which converts Update and Delete
expressions to UpdateSwap and DeleteSwap, respectively. Crucially, this
normalization rule also replaces the initial Scan of the mutation with a
single-row Values expression, which optimistically provides the complete
old row (if it exists and matches the filters). This allows the
UpdateSwap and DeleteSwap operators to use CPut to validate the
existence of the old row in the same batch as modifying it, changing the
mutation statement from 2 RTs to 1 RT.

With this normalization rule in place, UpdateSwap and DeleteSwap should
be available for use by any qualifying UPDATE and DELETE statements.

Fixes: #144503
Fixes: #85328

Release note (performance improvement): Add new `update swap` and
`delete swap` operators which allow some UPDATE and DELETE statements to
execute in 1 round trip instead of 2 round trips. For now, these
operators can be used when:

- all columns in the primary index are constrained to a single exact value
  by the WHERE clause;
- only a single row is modified;
- there are no FK checks or cascades;
- there are no uniqueness checks;
- there are no check constraints;
- there are no vector indexes modified;
- there are no passthrough columns to RETURNING;
- there are no triggers;
- the table only uses a single column family;
- there are no mutation columns or mutation indexes (i.e. the table is
  not undergoing an ALTER);
- there are no columns using composite encoding (e.g. DECIMAL, FLOAT,
  JSON, etc).

We may lift some of these restrictions in future releases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145019)
<!-- Reviewable:end -->

michae2 commented:
This has (at least) two problems that I need to figure out before it's ready:

1. The norm rules don't always apply when there are computed columns.
2. The norm rules don't apply at all when generic query plans are used.

michae2 commented:
Another problem:
3. suffers from the same issue as https://github.com/cockroachdb/cockroach/issues/145994#issuecomment-2877174901

michae2 commented:
> This has (at least) two problems that I need to figure out before it's ready:
> 
> 1. The norm rules don't always apply when there are computed columns.
> 2. The norm rules don't apply at all when generic query plans are used.
>
> Another problem: 3. suffers from the same issue as [#145994 (comment)](https://github.com/cockroachdb/cockroach/issues/145994#issuecomment-2877174901)

1. was fixed by adding the `UseSwapMutationWithProjectionProjection` rule
2. went away when (1) was fixed
3. was fixed by #147117 and #147510

So this should be ready now. There are a bunch of opttests. It needs more logictests, and benchmarking, and a session variable, but I think I'll do those in a follow-up PR since this is pretty big.

DrewKimball reviewed (commented):
Would it be useful to add a "must be any non-empty value" mode to `CPut`? If we had that, assuming all index keys were fully specified in the filters, we could make the following relaxations:
* (for update) If a column will be updated, don't require a filter on it, allow pruning it from the fetch columns, and allow it to have a composite type
* (for update) If a column is not in the RETURNING clause, don't require a filter on it, allow pruning it from the fetch columns, and allow it to have a composite type

> Reviewed 4 of 4 files at r1, 7 of 7 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-:-OU1kqSn84RjTsRM_1cv:b-4okjhi)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/delete` line 259 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1i6C41kyYTW4croVx:-OU1i6C41kyYTW4croVy:b-p5fs1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/exec/execbuilder/testdata/delete#L259)):*
> ```
> WHERE message LIKE '%Del%' OR message LIKE '%sending batch%'
> ----
> dist sender send  r74: sending batch 1 CPut, 1 EndTxn to (n1,s1):1
> ```

Looks like this test has rotted.
___
*[`pkg/sql/opt/norm/rules/mutation.opt` line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1hAtFDNtjHCkSTT0j:-OU1hAtFDNtjHCkSTT0k:b-fb9z9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/rules/mutation.opt#L70)):*
> ```
> #
> # This rule needs to run after PruneMutationInputCols and before
> # GenerateParameterizedJoin.
> ```

Note: `GenerateParameterizedJoin` is an exploration rule, so would never run before norm rules.
___
*[`pkg/sql/opt/norm/rules/mutation.opt` line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1hcM18bEe0oEBTvAg:-OU1eYO60V8tUnCZz8Iq:b-nidigb) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/rules/mutation.opt#L130)):*
> ```
> # This rule needs to run after PruneMutationInputCols and before
> # GenerateParameterizedJoin.
> [UseSwapMutationWithProjectionProjection, Normalize]
> ```

nit: up to you, but at this point I think it would be better to just match this pattern in go code, with handling for optional Project operators above the select + scan.
___
*[`pkg/sql/opt/norm/testdata/rules/mutation` line 395 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1jn1k9Aia7Hr3tvK_:-OU1jn1k9Aia7Hr3tvKa:bmdzozq) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/testdata/rules/mutation#L395)):*
> ```
> 
> # --------------------------------------------------
> # UseSwapMutation
> ```

Other things to test:
* Other types of indexes (inverted, vector, partial, expression)
* Hidden/inaccessible columns
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1pKIYAYjtOFIOJlOg:-OU1pKIYAYjtOFIOJlOh:b-hhuuu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/mutation_funcs.go#L326)):*
> ```Go
> // scalar expressions, the remaining filters, and a boolean indicating whether
> // all columns were constrained.
> func (c *CustomFuncs) constrainColsToScalarExprs(
> ```

It might be cleaner (and more complete) to use the constraints that are already calculated for filters via helpers like `memo.ExtractConstColumns` and `memo.ExtractValueForConstColumn`. You could do an early check to make sure that all needed columns are returned by `memo.ExtractConstColumns`, and then run `memo.ExtractValueForConstColumn` for each column individually in this function.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 337 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1WsvsC5MoEMlrXTtk:-OU1WsvsC5MoEMlrXTtl:bm44tj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/mutation_funcs.go#L337)):*
> ```Go
> 		if cols.Contains(col) {
> 			// Only use the first constraining filter found. If there is another
> 			// constraining filter, it will become part of requiredFilters.
> ```

nit: should it be `remainingFilters` instead of `requiredFilters`?
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 393 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1WMpR61XqS5Yh3zii:-OU1WMpR61XqS5Yh3zij:bo1ikls) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/mutation_funcs.go#L393)):*
> ```Go
> 				// `WHERE col IS NOT DISTINCT FROM <expr>` will be true if col has the
> 				// value <expr>, and <expr> is not a composite-encoded value.
> 				if !colinfo.CanHaveCompositeKeyEncoding(md.ColumnMeta(v.Col).Type) {
> ```

nit: could we instead check that none of the table's columns have a composite type in `CanUseSwapMutation`?


<!-- Sent from Reviewable.io -->

michae2 committed:
logictest: add testcase for incorrect dupe key from UpdateSwap

In #145994 we discovered a problem with batch ordering that could cause
the wrong ConditionFailed error to be returned. This could cause
UpdateSwap and DeleteSwap to incorrectly return a duplicate key error
instead of simply modifying 0 rows.

This problem was fixed by #147117 and #147510. Now that we have
UpdateSwap working, add a testcase that verifies the problem is fixed
for UpdateSwap.

Informs: #144503, #145994

Release note: None

michae2 committed:
sql, opt: add use_swap_mutations session variable

Release note (sql change): Add new session variable `use_swap_mutations`
which controls whether the new update swap and delete swap operators are
enabled for use by UPDATE and DELETE statements.

michae2 reviewed (commented):
Pushed with the following changes:
- rewrote it as a single rule with projection matching in go
- added more tests
- disabled these operators on tables with multiple column families. there are some bugs with multiple column families and so for now I want to wait
- added a session variable

> Would it be useful to add a "must be any non-empty value" mode to `CPut`?

Yes, definitely. I think this is what https://github.com/cockroachdb/cockroach/issues/71153 is getting at. For now I just wanted to do the more limited thing described in https://github.com/cockroachdb/cockroach/issues/144503 which doesn't require changes to the KV layer. But you're right, there's a more general blind update operation that is possible.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-:-OUsXkk-4JS79pFxVx9F:b6z9kef)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @rytaft)

___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1pKIYAYjtOFIOJlOg:-OUsa8Ls-q9clhwQzfKg:bm5y8mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/mutation_funcs.go#L326)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It might be cleaner (and more complete) to use the constraints that are already calculated for filters via helpers like `memo.ExtractConstColumns` and `memo.ExtractValueForConstColumn`. You could do an early check to make sure that all needed columns are returned by `memo.ExtractConstColumns`, and then run `memo.ExtractValueForConstColumn` for each column individually in this function.
</blockquote></details>

I looked at these pre-calculated constraints at first. But I don't think they exist when the RHS of the predicate is a more complicated scalar expression than just a constant. For example, if the predicate is something like `x = $1 * 2` I wouldn't expect one of these constraints to exist, but we can still safely build the swap mutation input using `$1 * 2` as the value for x.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 337 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1WsvsC5MoEMlrXTtk:-OUsYkL30HzKmHA01LZn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/mutation_funcs.go#L337)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: should it be `remainingFilters` instead of `requiredFilters`?
</blockquote></details>

Done.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 393 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1WMpR61XqS5Yh3zii:-OUsYmLe4BQ6w0Pp7Bsm:b-xm74yu) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/mutation_funcs.go#L393)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: could we instead check that none of the table's columns have a composite type in `CanUseSwapMutation`?
</blockquote></details>

I think this belongs here because it has to do with whether the scalar expressions are the only possible value for the LHS.

Oh, I wonder if instead of these type checks, I should be using `CanBeCompositeSensitive`? Probably...
___
*[`pkg/sql/opt/norm/rules/mutation.opt` line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1hAtFDNtjHCkSTT0j:-OUs_Pd76OdAEHOs9Tk_:bnigkid) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/rules/mutation.opt#L70)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Note: `GenerateParameterizedJoin` is an exploration rule, so would never run before norm rules.
</blockquote></details>

Yeah, just wanted to note the relationship.
___
*[`pkg/sql/opt/norm/rules/mutation.opt` line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1hcM18bEe0oEBTvAg:-OUsZF9P68FVs5UhblNM:b561sin) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/rules/mutation.opt#L130)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: up to you, but at this point I think it would be better to just match this pattern in go code, with handling for optional Project operators above the select + scan.
</blockquote></details>

Great point! This made it much simpler. Now there's just a single rule.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/delete` line 259 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1i6C41kyYTW4croVx:-OUsa4xU0aPf21JWGwlJ:b1oxlv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/exec/execbuilder/testdata/delete#L259)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Looks like this test has rotted.
</blockquote></details>

Ah drat, good catch.
___
*[`pkg/sql/opt/norm/testdata/rules/mutation` line 395 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1jn1k9Aia7Hr3tvK_:-OUsZP9r6QpSuOKr5y-5:b-pb99zx) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/testdata/rules/mutation#L395)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Other things to test:
* Other types of indexes (inverted, vector, partial, expression)
* Hidden/inaccessible columns
</blockquote></details>

Good call. I added a bunch of tests. (And TODOs for things that don't work yet.)

Inverted, partial, and expression indexes all seem to work fine. Vector indexes do not work yet because of the `VectorMutationSearchExpr` operators. I added a TODO.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> Reviewed 11 of 11 files at r4, 12 of 12 files at r5, 2 of 2 files at r6, 9 of 9 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145019)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @rytaft)

___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1pKIYAYjtOFIOJlOg:-OUvVoSjAcJPsiK_tX19:b-m9qjsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/mutation_funcs.go#L326)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I looked at these pre-calculated constraints at first. But I don't think they exist when the RHS of the predicate is a more complicated scalar expression than just a constant. For example, if the predicate is something like `x = $1 * 2` I wouldn't expect one of these constraints to exist, but we can still safely build the swap mutation input using `$1 * 2` as the value for x.
</blockquote></details>

Hm, nevermind, then. Thanks for checking.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 393 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1WMpR61XqS5Yh3zii:-OUvWALI0EGuGubE0ndd:b-2v3y8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/mutation_funcs.go#L393)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I think this belongs here because it has to do with whether the scalar expressions are the only possible value for the LHS.

Oh, I wonder if instead of these type checks, I should be using `CanBeCompositeSensitive`? Probably...
</blockquote></details>

`CanBeCompositeSensitive` is just used to determine whether the result of an expression can change when the value of a variable it depends on is changed to an equivalent composite value, e.g. from `1.0` to `1.00` for a decimal. We're not remapping the expression to use different, equivalent columns or anything, so I don't think it's a concern. We're just concerned about the case where the row being updated has different (but equivalent) composite values vs the filters.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/delete` line 259 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1i6C41kyYTW4croVx:-OUvYmu-A19u1AsLMzK5:b-943cpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/exec/execbuilder/testdata/delete#L259)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Ah drat, good catch.
</blockquote></details>

Would it work to just change the column to a decimal?
___
*[`pkg/sql/opt/norm/testdata/rules/mutation` line 1888 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUviIYAEzDc-vrdt9Zk:-OUviIYAEzDc-vrdt9Zl:bs78m76) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f116915d55c745a549189b0f9dd7dddd27ee1b7/pkg/sql/opt/norm/testdata/rules/mutation#L1888)):*
> ```
>                           └── c:14 != cd.c:10 [outer=(10,14), constraints=(/10: (/NULL - ]; /14: (/NULL - ])]
> 
> # Test that we do use swap mutations with before triggers.
> ```

Hm I didn't realize we'd allow BEFORE triggers. I don't think it'll be correct when there is no matching row, for example:
```
root@localhost:26257/defaultdb> CREATE TABLE t (x INT PRIMARY KEY, y INT);
                             ->
                             -> CREATE FUNCTION f() RETURNS TRIGGER LANGUAGE PLpgSQL AS $$ BEGIN RAISE NOTICE '% -> %', OLD, NEW; RETURN
                             -> COALESCE(NEW, OLD); END; $$;
                             ->
                             -> CREATE TRIGGER t BEFORE UPDATE OR DELETE ON t FOR EACH ROW EXECUTE FUNCTION f();
CREATE TABLE

Time: 42ms total (execution 22ms / network 21ms)

CREATE FUNCTION

Time: 53ms total (execution 29ms / network 24ms)

CREATE TRIGGER

Time: 255ms total (execution 255ms / network 0ms)

root@localhost:26257/defaultdb> UPDATE t SET x = x + 100 WHERE x = 1 AND y = 2;
UPDATE 0

Time: 6ms total (execution 5ms / network 0ms)

root@localhost:26257/defaultdb> DELETE FROM t WHERE x = 1 AND y = 2;
DELETE 0

Time: 4ms total (execution 4ms / network 0ms)
```
Notice there are no notices printed, since the BEFORE trigger fires only on existing rows that will be updated or deleted.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 289 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvUYSYEm8nwndLfrqo:-OUvUYSYEm8nwndLfrqp:b-avibw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/norm/mutation_funcs.go#L289)):*
> ```Go
> 			}
> 		}
> 		// TODO: case *memo.VectorMutationSearchExpr
> ```

You could make the argument that it'll be better not to use this when there's a vector index, since searching the index is expensive.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 318 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvV20eENaVmvZfYpRA:-OUvV20eENaVmvZfYpRB:bdvywcc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/norm/mutation_funcs.go#L318)):*
> ```Go
> 			newRow = append(newRow, colExpr)
> 			newTypes = append(newTypes, md.ColumnMeta(col).Type)
> 			// can we steal the col ID like this?
> ```

Ooh, that's a good point. These columns will have been added to the metadata with the understanding that they originate from the mutated table; this implies that the values in the column are a subset of the values in the table, possibly with nulls added e.g. by outer joins. There are places where we depend on that assumption elsewhere in the optimizer, so we have to be careful about re-using column IDs in a way that can "add new values".

Unfortunately, I think we need to be synthesizing new columns, and replacing the old ones in whatever expressions reference them, as well as the mutation's fetch columns list.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/swap_mutation` line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvaYi-0W-3JSCO8KoJ:-OUvaYi-0W-3JSCO8KoK:b-64e4pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/exec/execbuilder/testdata/swap_mutation#L130)):*
> ```
> ----
> 
> # Test swap mutations with computed columns.
> ```

Maybe it's not that important, but what would you expect to happen in the following test case?
```
CREATE TABLE ab (
  a INT PRIMARY KEY,
  b FLOAT AS ((1 / a::FLOAT)) STORED
);
UPDATE ab SET a = 0 WHERE a = 1 AND b IS NULL;
```
We'll see side effects from the computed column expressions even though the row to be updated doesn't exist. I'm all for saying this is totally fine according to our [volatility policy](https://github.com/cockroachdb/cockroach/blob/691f7a802d9b7be72117ed5251df915fa638f628/pkg/sql/opt/props/volatility.go#L25), though.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/swap_mutation` line 150 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvc2obERszjJ9RAyfm:-OUvc2obERszjJ9RAyfn:b-2bw8nv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/exec/execbuilder/testdata/swap_mutation#L150)):*
> ```
> 
> query T
> EXPLAIN UPDATE abcde SET b = 0 WHERE a = 1 AND b = 1 AND e = 1
> ```

In the future we could also drop the requirement for the `e = 1` filter, right?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/swap_mutation` line 371 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvd90j575jpURJ2mFL:-OUvd90j575jpURJ2mFM:b-f1thur) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/exec/execbuilder/testdata/swap_mutation#L371)):*
> ```
> 
> query T kvtrace
> UPDATE mno SET o = 'o' || o WHERE m = 'm1' AND n = 'n1' AND o = 'o'
> ```

Can we do a similar test where `m` is being modified by the update instead? I'd expect to see the partial index being updated in that case.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/swap_mutation` line 528 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvehi8447LLbvb5b5g:-OUvehi8447LLbvb5b5h:b-9wx7f8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/exec/execbuilder/testdata/swap_mutation#L528)):*
> ```
>   v INT,
>   PRIMARY KEY (r, s),
>   INDEX (t) STORING (u, v),
> ```

It'd be nice to have some STORING tests in the case when a swap mutation can be used.


<!-- Sent from Reviewable.io -->

michae2 commented:
I pushed two WIP commits trying to fix a couple of problems, but there are some bugs that I still need to work out.

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145019)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @rytaft)

___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 318 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvV20eENaVmvZfYpRA:-OV2s9k78ZmVwQQ5RTGT:bikctbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/norm/mutation_funcs.go#L318)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Ooh, that's a good point. These columns will have been added to the metadata with the understanding that they originate from the mutated table; this implies that the values in the column are a subset of the values in the table, possibly with nulls added e.g. by outer joins. There are places where we depend on that assumption elsewhere in the optimizer, so we have to be careful about re-using column IDs in a way that can "add new values".

Unfortunately, I think we need to be synthesizing new columns, and replacing the old ones in whatever expressions reference them, as well as the mutation's fetch columns list.
</blockquote></details>

I tried synthesizing new columns in the last WIP commit, but it's tricky to get right. It's currently not replacing the old columns in the mutation's fetch column list (which is why I've marked this do-not-merge). And if the mutation has a RETURNING clause then it seems these column IDs could escape "upward" out of the mutation and need to be replaced outside the scope of the normalization rule... I'm not sure how to deal with that. It seems like the replacement needs to be "global" for the entire plan but the normalization rule only has a "local" power over part of the tree.

Could we keep the same column IDs but change the metadata? Or could we prohibit the parts of the optimizer that depend on this assumption (what parts are those)?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/swap_mutation` line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvaYi-0W-3JSCO8KoJ:-OV2teMBA-E_Be8QuDAQ:b7j3udi) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/exec/execbuilder/testdata/swap_mutation#L130)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Maybe it's not that important, but what would you expect to happen in the following test case?
```
CREATE TABLE ab (
  a INT PRIMARY KEY,
  b FLOAT AS ((1 / a::FLOAT)) STORED
);
UPDATE ab SET a = 0 WHERE a = 1 AND b IS NULL;
```
We'll see side effects from the computed column expressions even though the row to be updated doesn't exist. I'm all for saying this is totally fine according to our [volatility policy](https://github.com/cockroachdb/cockroach/blob/691f7a802d9b7be72117ed5251df915fa638f628/pkg/sql/opt/props/volatility.go#L25), though.
</blockquote></details>

Good point. We could bail out if any of the computed column expressions are not Leakproof.
___
*[`pkg/sql/opt/norm/testdata/rules/mutation` line 1888 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUviIYAEzDc-vrdt9Zk:-OV2ua_YBQ7KGBhqucFY:b-be6qm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f116915d55c745a549189b0f9dd7dddd27ee1b7/pkg/sql/opt/norm/testdata/rules/mutation#L1888)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Hm I didn't realize we'd allow BEFORE triggers. I don't think it'll be correct when there is no matching row, for example:
```
root@localhost:26257/defaultdb> CREATE TABLE t (x INT PRIMARY KEY, y INT);
                             ->
                             -> CREATE FUNCTION f() RETURNS TRIGGER LANGUAGE PLpgSQL AS $$ BEGIN RAISE NOTICE '% -> %', OLD, NEW; RETURN
                             -> COALESCE(NEW, OLD); END; $$;
                             ->
                             -> CREATE TRIGGER t BEFORE UPDATE OR DELETE ON t FOR EACH ROW EXECUTE FUNCTION f();
CREATE TABLE

Time: 42ms total (execution 22ms / network 21ms)

CREATE FUNCTION

Time: 53ms total (execution 29ms / network 24ms)

CREATE TRIGGER

Time: 255ms total (execution 255ms / network 0ms)

root@localhost:26257/defaultdb> UPDATE t SET x = x + 100 WHERE x = 1 AND y = 2;
UPDATE 0

Time: 6ms total (execution 5ms / network 0ms)

root@localhost:26257/defaultdb> DELETE FROM t WHERE x = 1 AND y = 2;
DELETE 0

Time: 4ms total (execution 4ms / network 0ms)
```
Notice there are no notices printed, since the BEFORE trigger fires only on existing rows that will be updated or deleted.
</blockquote></details>

Ah, great point. So we need to explicitly disallow before triggers, too. (I think they're implicitly disallowed right now due to Barriers showing up in the plan, but we should make it explicit.)


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145019)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @rytaft)

___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 318 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvV20eENaVmvZfYpRA:-OVEM_pUDP2ZsU1j2R86:b7uyhgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/norm/mutation_funcs.go#L318)):*
>And if the mutation has a RETURNING clause then it seems these column IDs could escape "upward" out of the mutation and need to be replaced outside the scope of the normalization rule

This at least I don't think is too problematic in theory - we could remap back to the original column IDs after the mutation, since at that point we've filtered out nonexistent values. That would just amount to adding a Project operator on top of the mutation after making the replacement.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Pushed a revision with the following changes:
- now using new column IDs for the output columns of the ValuesExpr, and remapping scalar expressions above
- now explicitly disallowing all triggers, vector indexes, mutation columns, mutation indexes
- explicitly disallowing multi-column-family tables due to a bug / not enough testing
- checking that scalar expression types are identical to column type

Probably needs more tests still, but it feels like it's getting closer. I think this is RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-:-OXAJnKQAA5kM2EYvlGq:b-uei1k3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @rytaft)

___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 393 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1WMpR61XqS5Yh3zii:-OXAI6SW8hhhbZIOfTcJ:b1l6nhf) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/mutation_funcs.go#L393)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

`CanBeCompositeSensitive` is just used to determine whether the result of an expression can change when the value of a variable it depends on is changed to an equivalent composite value, e.g. from `1.0` to `1.00` for a decimal. We're not remapping the expression to use different, equivalent columns or anything, so I don't think it's a concern. We're just concerned about the case where the row being updated has different (but equivalent) composite values vs the filters.
</blockquote></details>

Thanks for explaining. I'm sticking with checking `colinfo.CanHaveCompositeKeyEncoding` of the column types, and I've moved the check into the `useExprAsScalarConstraint` helper function. I've also added a check that the scalar expression type is identical to the column type.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 289 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvUYSYEm8nwndLfrqo:-OXAIjnhApclxgepQmI6:bnnaow1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/norm/mutation_funcs.go#L289)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

You could make the argument that it'll be better not to use this when there's a vector index, since searching the index is expensive.
</blockquote></details>

Good point. I've added an explicit disallow when there's a vector index.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 318 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvV20eENaVmvZfYpRA:-OXAIx1pCqB8YdZ_WO85:b-xzlxwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/norm/mutation_funcs.go#L318)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

>And if the mutation has a RETURNING clause then it seems these column IDs could escape "upward" out of the mutation and need to be replaced outside the scope of the normalization rule

This at least I don't think is too problematic in theory - we could remap back to the original column IDs after the mutation, since at that point we've filtered out nonexistent values. That would just amount to adding a Project operator on top of the mutation after making the replacement.
</blockquote></details>

That makes sense, thank you. I'm now generating new column IDs in `buildSwapMutationValues` and remapping all of the scalar expressions above. As it turns out, when we get to the mutation, it already remaps all column IDs *except* passthrough columns. For now I've simply disallowed swap mutations when there are passthrough columns.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/delete` line 259 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1i6C41kyYTW4croVx:-OXAJPfE4-JrEar6Dffs:b-duzahh) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/exec/execbuilder/testdata/delete#L259)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Would it work to just change the column to a decimal?
</blockquote></details>

Yes! Done.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/swap_mutation` line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvaYi-0W-3JSCO8KoJ:-OXAJWupEX65Quw2USuC:b6jbd1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/exec/execbuilder/testdata/swap_mutation#L130)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Good point. We could bail out if any of the computed column expressions are not Leakproof.
</blockquote></details>

I decided not to do anything about this. As you point out, it seems like this is in line with our volatility policy.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/swap_mutation` line 150 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUvc2obERszjJ9RAyfm:-OXAJTIFDcdTL1HOjaqJ:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d512eb236d03b08fcefefce51344d28c4fdcc01/pkg/sql/opt/exec/execbuilder/testdata/swap_mutation#L150)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

In the future we could also drop the requirement for the `e = 1` filter, right?
</blockquote></details>

Yes.
___
*[`pkg/sql/opt/norm/testdata/rules/mutation` line 1888 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OUviIYAEzDc-vrdt9Zk:-OXAJg5mAsMWWrrvs8Is:b-qynl48) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f116915d55c745a549189b0f9dd7dddd27ee1b7/pkg/sql/opt/norm/testdata/rules/mutation#L1888)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Ah, great point. So we need to explicitly disallow before triggers, too. (I think they're implicitly disallowed right now due to Barriers showing up in the plan, but we should make it explicit.)
</blockquote></details>

I've explicitly disallowed all triggers (before, after, insteadof).


<!-- Sent from Reviewable.io -->

rytaft reviewed (dismissed):
Nice work!

> @rytaft reviewed 1 of 4 files at r1, 2 of 11 files at r4, 21 of 21 files at r18, 13 of 13 files at r19, 2 of 2 files at r20, 10 of 10 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-:-OYIIqJF2BQK4EFTMi6c:bdtx8r6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @michae2)

___
*[`-- commits` line 18 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYH_curDNDFr3_EStwm:-OYH_curDNDFr3_EStwn:b-bg6qrc):*
nit: replace -> replaces
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 244 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYHaeUoE-kutwe3lnnc:-OYHaeUoE-kutwe3lnnd:b-qvlns) ([raw file](https://github.com/cockroachdb/cockroach/blob/c259431c954bcf8d7be59446b9ee77e5d32c7a70/pkg/sql/opt/norm/mutation_funcs.go#L244)):*
> ```Go
> 
> 	// Passthrough columns are not currently remapped correctly by UseSwapMutation
> 	// after BuildSwapMutationInput generates new column IDs.
> ```

This isn't one of the cases you list as being unsupported in the comment / release note for update / delete swap
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 253 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYHjMa46TKkyvLS-573:-OYHjMa46TKkyvLS-574:bea5r9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/c259431c954bcf8d7be59446b9ee77e5d32c7a70/pkg/sql/opt/norm/mutation_funcs.go#L253)):*
> ```Go
> 		return false
> 	}
> ```

should you also check for UniqueWithTombstoneIndexes here? How will we know about uniqueness checks for serializable queries?
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 265 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYHay4UAkO9XbL7OEPF:-OYHay4UAkO9XbL7OEPG:bl78aje) ([raw file](https://github.com/cockroachdb/cockroach/blob/c259431c954bcf8d7be59446b9ee77e5d32c7a70/pkg/sql/opt/norm/mutation_funcs.go#L265)):*
> ```Go
> 	}
> 	beforeTriggers := cat.GetRowLevelTriggers(table, tree.TriggerActionTimeBefore, eventsToMatch)
> 	if len(beforeTriggers) != 0 {
> ```

you only list after triggers as being unsupported in the comment / release note
___
*[`pkg/sql/opt/exec/execbuilder/testdata/swap_mutation` line 649 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYIGuevDufcw3_oh9UT:-OYIGuevDufcw3_oh9UU:b1z7kht) ([raw file](https://github.com/cockroachdb/cockroach/blob/c259431c954bcf8d7be59446b9ee77e5d32c7a70/pkg/sql/opt/exec/execbuilder/testdata/swap_mutation#L649)):*
> ```
> ----
> 
> # Test swap mutations with a single column family.
> ```

are you testing single or multiple families now? looks like still multiple (maybe this is duplicated from above?)
___
*[`pkg/sql/opt/norm/testdata/rules/mutation` line 529 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYIHUc18MIRpdmz185s:-OYIHUc18MIRpdmz185t:bdoqrs4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c259431c954bcf8d7be59446b9ee77e5d32c7a70/pkg/sql/opt/norm/testdata/rules/mutation#L529)):*
> ```
>       └── (1, NULL, NULL)
> 
> norm expect-not=UseSwapMutation
> ```

can you add a comment here and for the other expect-not cases for why you don't expect this rule?
___
*[`pkg/sql/opt/norm/testdata/rules/mutation` line 809 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYIHdiX4cB-gjvVYLPg:-OYIHdiX4cB-gjvVYLPh:b7se8gf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c259431c954bcf8d7be59446b9ee77e5d32c7a70/pkg/sql/opt/norm/testdata/rules/mutation#L809)):*
> ```
> 
> # Test that we do not use swap mutations when there is a column with a composite
> # key encoding (such as DECIMAL).
> ```

this is not listed as a limitation in the comment/release note
___
*[`pkg/sql/opt/norm/prune_cols_funcs.go` line 105 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYIMOb8CySwkiMP0ptu:-OYIMOb8CySwkiMP0ptv:bpk371p) ([raw file](https://github.com/cockroachdb/cockroach/blob/effc4cfbf2e1eebb26dabf6f42cd7bc865084457/pkg/sql/opt/norm/prune_cols_funcs.go#L105)):*
> ```Go
> 
> 	// familyCols returns the columns in the given family.
> 	// TODO combine this with helper func
> ```

looks like a leftover?
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 337 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYIJPP020NQ2iBt676f:-OYIJPP19hOv01Bu1Rce:b-qu9vkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/effc4cfbf2e1eebb26dabf6f42cd7bc865084457/pkg/sql/opt/norm/mutation_funcs.go#L337)):*
> ```Go
> 	case *memo.SelectExpr:
> 		// Swap mutations can only operate on a single row.
> 		if c.HasZeroOrOneRow(t) {
> ```

Aren't you already checking this in the rule itself?
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 419 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYILAG1CfII-SDSgcnz:-OYILAG1CfII-SDSgco-:b-9ijg6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/effc4cfbf2e1eebb26dabf6f42cd7bc865084457/pkg/sql/opt/norm/mutation_funcs.go#L419)):*
> ```Go
> // scalar expressions, the remaining filters, and a boolean indicating whether
> // all columns were constrained.
> func (c *CustomFuncs) constrainColsToScalarExprs(
> ```

Is there no existing logic we can reuse for this? Seems like there's a lot of stuff in here that's hard to verify without a lot of testing.


<!-- Sent from Reviewable.io -->

yuzefovich commented on pkg/sql/opt/exec/execbuilder/testdata/delete:
```diff
@@ -221,7 +221,7 @@ vectorized: true
-CREATE TABLE a (a INT PRIMARY KEY)
+CREATE TABLE a (a DECIMAL PRIMARY KEY)
 
 # Delete range operates in chunks of 600 (defined by row.TableTruncateChunkSize).
```
nit: drive-by that we should do `s/TableTruncateChunkSize/deleteRangeChunkSize/`.

michae2 committed:
memo: add ExtractExactScalarExprsForColumns

Add helper function `memo.ExtractExactScalarExprsForColumns`. This
function is similar to the constraint builder in that it constrains
columns using FiltersExpr, but it differs slightly in details. Unlike
constraint builder, this function:
- constrains columns to scalar expressions instead of constants
- always constrains tightly
- in fact, always constrains _exactly_ (that is, to scalar expressions that
   produce not only equivalent values, but _exactly_ the same bytes)

These differences are necessary for UpdateSwap and DeleteSwap, because
for swap mutations we need (a) the ability to use placeholders and
casting and other scalar expressions to feed the swap mutations, and (b)
exact byte equivalence for CPut to work correctly.

Informs: #144503
Informs: #85328

Release note: None

michae2 reviewed (commented):
Pushed a revision with the following changes:
- checking for `UniqueWithTombstoneIndexes`
- moved `constrainColsToScalarExprs` to `extract.go`, renamed it to `ExtractExactScalarExprsForColumns` and pulled it into a separate commit
- various other review comments

If we like where this is going, I can work on tests tomorrow.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-:-O_pjVJoA3ayY20hRsd2:b-c083ld)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @rytaft)

___
*[`-- commits` line 18 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYH_curDNDFr3_EStwm:-O_pjaio0G-GLHa3IUd7:b-896fix):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: replace -> replaces
</blockquote></details>

Done.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OU1pKIYAYjtOFIOJlOg:-O_pjdp5EVkG7eP3Bfwt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59a7cadf9fd6493e3060ddbab3b76bdef5ff78b7/pkg/sql/opt/norm/mutation_funcs.go#L326)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Hm, nevermind, then. Thanks for checking.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 244 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYHaeUoE-kutwe3lnnc:-O_pjfhB3VM7XbiqOa0T:b-m9ym7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c259431c954bcf8d7be59446b9ee77e5d32c7a70/pkg/sql/opt/norm/mutation_funcs.go#L244)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This isn't one of the cases you list as being unsupported in the comment / release note for update / delete swap
</blockquote></details>

Added it to the comment / release note.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 253 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYHjMa46TKkyvLS-573:-O_pjlL20wYS2g8s02p2:b-aa9bn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c259431c954bcf8d7be59446b9ee77e5d32c7a70/pkg/sql/opt/norm/mutation_funcs.go#L253)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should you also check for UniqueWithTombstoneIndexes here? How will we know about uniqueness checks for serializable queries?
</blockquote></details>

Good catch! I added a check for UniqueWithTombstoneIndexes.

Uniqueness checks for serializable (and also FK checks) are handled by `[]` in the body of the UseSwapMutation rule.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 265 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYHay4UAkO9XbL7OEPF:-O_pkQa17p2pgcGSfyj7:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c259431c954bcf8d7be59446b9ee77e5d32c7a70/pkg/sql/opt/norm/mutation_funcs.go#L265)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

you only list after triggers as being unsupported in the comment / release note
</blockquote></details>

Fixed.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 337 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYIJPP020NQ2iBt676f:-O_pkT9t1elaJpmhLTb2:bs2vhuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/effc4cfbf2e1eebb26dabf6f42cd7bc865084457/pkg/sql/opt/norm/mutation_funcs.go#L337)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Aren't you already checking this in the rule itself?
</blockquote></details>

Good point, removed.
___
*[`pkg/sql/opt/norm/mutation_funcs.go` line 419 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYILAG1CfII-SDSgcnz:-O_pklpF0VCuk3fzDspV:b-r8kjm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/effc4cfbf2e1eebb26dabf6f42cd7bc865084457/pkg/sql/opt/norm/mutation_funcs.go#L419)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is there no existing logic we can reuse for this? Seems like there's a lot of stuff in here that's hard to verify without a lot of testing.
</blockquote></details>

Drew asked something similar above. I think the closest existing logic is constraint builder, but this logic needs to be a little different to work for swap mutations.
- This logic needs to constrain to _scalar expressions_ instead of just constants, in order to handle placeholders and casting and computed columns.
- This logic always needs to constrain tightly.
- And furthermore this logic always needs to constrain _exactly_ (that is, to scalar expressions that produce not only equivalent values, but the exact same bytes).

(I added this info to some comments and to the commit message of the second commit.)
___
*[`pkg/sql/opt/norm/prune_cols_funcs.go` line 105 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYIMOb8CySwkiMP0ptu:-O_pmSaJ-8QgK9GINj0M:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/effc4cfbf2e1eebb26dabf6f42cd7bc865084457/pkg/sql/opt/norm/prune_cols_funcs.go#L105)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

looks like a leftover?
</blockquote></details>

Removed.
___
*[`pkg/sql/opt/norm/testdata/rules/mutation` line 809 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-OYIHdiX4cB-gjvVYLPg:-O_pm_uk8tBdaUR_rta6:b-m9ym7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c259431c954bcf8d7be59446b9ee77e5d32c7a70/pkg/sql/opt/norm/testdata/rules/mutation#L809)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

this is not listed as a limitation in the comment/release note
</blockquote></details>

Added it to the comment / release note.


<!-- Sent from Reviewable.io -->

DrewKimball commented on pkg/sql/opt/memo/extract.go:
```diff
@@ -540,3 +543,126 @@ func ExtractTailCalls(expr opt.Expr, tailCalls map[opt.ScalarExpr]struct{}) {
+		colType := md.ColumnMeta(col).Type
+		if colinfo.CanHaveCompositeKeyEncoding(colType) {
+			// TODO(michae2): If ExtractConstDatum(e) returns a non-composite datum,
+			// we could still constrain using this expression.
```
Is this true? Could we have the value in the table be `1.000::DECIMAL` and the one in the filter `1::DECIMAL`?

DrewKimball commented on pkg/sql/opt/memo/extract.go:
```diff
@@ -540,3 +543,126 @@ func ExtractTailCalls(expr opt.Expr, tailCalls map[opt.ScalarExpr]struct{}) {
+// references more than one of the given columns, the filter will be added to
+// remainingFilters. ExtractExactScalarExprsForColumns returns the mapping from
+// constrained columns to scalar expressions, the remaining filters, and a
+// boolean indicating whether all columns were constrained.
```
nit: It may be worth noting that the returned remaining filters can include some synthesized filters.

DrewKimball replied:
Bump. Also just a comment, so not blocking.

DrewKimball reviewed (commented):
>If we like where this is going, I can work on tests tomorrow.

I like it!

rytaft reviewed (commented):
Me too :) 

> @rytaft reviewed 1 of 1 files at r23, 8 of 12 files at r24, 11 of 11 files at r26, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145019#-:-Oa_LWPA4o3AqxYszGRE:bplp5r6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @michae2)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
Nice work! LGTM

michae2 commented:
TFTRs! I'm going to merge before I forget, and will fix those comments in the next PR.

bors r=DrewKimball

yuzefovich commented:
Dismissed Becca's stale review.

bors r=drewkimball

michae2 commented:
bors r=DrewKimball
